solidity/contracts/relay/Relay.sol

```diff
@@ -0,0 +1,507 @@
+// SPDX-License-Identifier: MIT
+
+// ██████████████     ▐████▌     ██████████████
+// ██████████████     ▐████▌     ██████████████
+//               ▐████▌    ▐████▌
+//               ▐████▌    ▐████▌
+// ██████████████     ▐████▌     ██████████████
+// ██████████████     ▐████▌     ██████████████
+//               ▐████▌    ▐████▌
+//               ▐████▌    ▐████▌
+//               ▐████▌    ▐████▌
+//               ▐████▌    ▐████▌
+//               ▐████▌    ▐████▌
+//               ▐████▌    ▐████▌
+
+pragma solidity ^0.8.9;
+
+import "@openzeppelin/contracts/access/Ownable.sol";
+
+import {BytesLib} from "@keep-network/bitcoin-spv-sol/contracts/BytesLib.sol";
+import {BTCUtils} from "@keep-network/bitcoin-spv-sol/contracts/BTCUtils.sol";
+import {ValidateSPV} from "@keep-network/bitcoin-spv-sol/contracts/ValidateSPV.sol";
+
+import "../bridge/IRelay.sol";
+
+struct Epoch {
+    uint32 timestamp;
+    // By definition, bitcoin targets have at least 32 leading zero bits.
+    // Thus we can only store the bits that aren't guaranteed to be 0.
+    uint224 target;
+}
+
+library RelayUtils {
+    using BytesLib for bytes;
+
+    /// @notice Extract the timestamp of the header at the given position.
+    /// @param headers Byte array containing the header of interest.
+    /// @param at The start of the header in the array.
+    /// @return The timestamp of the header.
+    /// @dev Assumes that the specified position contains a valid header.
+    /// Performs no validation whatsoever.
+    function extractTimestampAt(bytes memory headers, uint256 at)
+        internal
+        pure
+        returns (uint32)
+    {
+        return BTCUtils.reverseUint32(uint32(headers.slice4(68 + at)));
+    }
+}
+
+interface ILightRelay is IRelay {
+    event Genesis(uint256 blockHeight);
+    event Retarget(uint256 oldDifficulty, uint256 newDifficulty);
+    event ProofLengthChanged(uint256 newLength);
+
+    function retarget(bytes memory headers) external;
+
+    function validateChain(bytes memory headers)
+        external
+        view
+        returns (uint256);
+
+    function getBlockDifficulty(uint256 blockNumber)
+        external
+        view
+        returns (uint256);
+
+    function getEpochDifficulty(uint256 epochNumber)
+        external
+        view
+        returns (uint256);
+
+    function getRelayRange()
+        external
+        view
+        returns (uint256 relayGenesis, uint256 currentEpochEnd);
+}
+
+contract Relay is Ownable, ILightRelay {
+    using BytesLib for bytes;
+    using BTCUtils for bytes;
+    using ValidateSPV for bytes;
+    using RelayUtils for bytes;
+
+    bool public ready;
+    // Number of blocks required for a retarget proof.
+    // Governable
+    // Should be set to a fairly high number (e.g. 20-50) in production.
```

Comment:
Something we need to be aware of and something we could make explicit in the documentation is that the higher  `proofLength`, the longer delay with submitting proofs for transactions made just before the epoch ends.

That is, I need to wait for enough headers in the new epoch to accumulate so that the relay can `retarget` and I can prove the work in `validateChain`.

For 50 blocks, it means ~8h delay assuming 6 blocks are mined per hour.